### DISCUSSION TITLE
Player SetIndex problem when using I2s

### DISCUSSION THREAD

**Discussion by waszee** [pschatzmann/arduino-audio-tools#1774] [Q&A] [answered]
I am having a problem changing internet radio url stations from an I2s output setup.  I am trying to use an IR remote to change the index number and then pass the new index number to player.setIndex(idx).  On my system the player appears to be trying to switch the metadata stream to the correct index,  but the audio output is not changing from station 0.  Similar code worked when using a Thinker board with the audiokit so think the problem is with my i2s implementation. If I use player.next() and player.previous() functions to move up or down a single step those functions work.  The Ir control  of volume works well.  I am just having trouble with the setIndex function.  Would like suggestions as I have a tried several variants including trying to change at the source without success and I see no problems in my look at the library codes.  Audio output does play the startup station and that station resumes playing after trying to make a change.

"""
```
void loop(){
  player.copy();

  thestation=dstation;
  thevol=dvol;
  if (irrecv.decode(&results)) {
      check_vol_change();
      check_station_change();
      irrecv.resume();  // Receive the next value
  }
  if (dvol!=thevol){
    player.setVolume(dvol);;
  }
  if (dstation!=thestation) {
    player.end();
    player.setIndex(dstation);
    Serial.printf("station = %d  vol = %f \n", dstation,dvol);
    delay(1000);
    player.begin();
    
    //player.setActive(true);
    player.copy();
  }
    
}
```

**Comment by pschatzmann** [comment, marked as answer]
Look up in the class documentation what begin is doing !
The system does exactly what you tell it to do...

**Comment by waszee** [comment]
I found the hint "start" confusing as player does not have a start function.  The example codes all use player.next() which works.  I did not find an example uising player.setIndex(idx) or player.setStream(Stream * stream).  The following code is switching stations, gives an error message about icy-metaint not defined but then seems to print out the station and audio song information so error needs handling but not sure what the problem.

""""
`  if (dstation!=thestation) {
    player.stop();
    urlStream.begin(urls[dstation]);
    player.setIndex(dstation);
    Serial.printf("station = %d  vol = %f \n", dstation,dvol);
    delay(1000);
    player.play();
    player.copy();
  }

'''The image is clip of the serial monitor output showing the error and warning messages.
![image](https://github.com/user-attachments/assets/7a35a71a-9c0d-472f-9bcc-09a5033b5353)

**Comment by pschatzmann** [comment]
You were supposed to read the [class documentation](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html#aa83ff9b2ca349dc7352f9904d883a5e0) and not search for examples!
Why are you still calling setindex and not directly urlStream.begin(dstation); ?
Also your player.play is absolutely unnecessary, since you called the begin with an implicit isActive true and you were setting the player already active with this.

Also are you sure that you are connnecting to an Icecast/Shoutcast Server ? I think it is not, so the system just let's you know.
If you don't need icecast metadata, just use a URLStream instead....

**Comment by waszee** [comment]
Did read the class documentation but did not comprehend "start" so started to look for an example.  I did try to drop the player.setIndex line as thought unnecessary.  The station changed but I did not get any audio song metadata from the stations after the station change.  I added the line back and the metadata started to stream in addition to the audio out.  I  am getting the Error message:
E] MetaDataCY.h:269 - icy metaint not defined 

But after a moment or two the station metadata streams out followed by Title song information steams when new song.  This is what I wanted.   FYI  urlSteam   was defined as as an ICYStream type.  I am not unhappy at this point but do not understand the error message.  I wonder if I should add a flush statement but thought the player.stop or player.end would have handled the left overs in the buffer from the previous station and am not surprised that changing to a new station might not detect the metadata on startup.  

I do have a couple more issues with changing and playing stations.  Stations with poor or slow steam often include momentary pause in the play.  I am using the default buffer but have wondered if changing it might help.  It looked like I could set buffer size using player.copy(buffer) option but I saw several other ways in the documentation and not sure what is preferred.  I assume it should change by power of 2 Correct, i.e. 512, 1024, 2048 etc..  Another issue is a failed connection.  I think the error code is 223.  The timeout seems too long to me.  I think it is 60 seconds and the IR remote is just frozen until that timeout passes.  Can I use a define to override the default wait?

**Comment by pschatzmann** [comment]
You did not provide any url which gives this error. But you can test them yourself with wget;

    wget --debug --header="Icy-MetaData: 1" http://stream.srg-ssr.ch/m/rsj/mp3_128

You should get a response that contains an icy-metaint line. E.g 

   icy-metaint:16000

Maybe you must call end() to close the connection properly before opening a new one.

I think you talk about the amount of data that is copied to the decoder with each copy command, and that should not make any difference. You can try the buffered implementation by changing the class, but I never had any need to do this.

You can change the timeouts via the API: please check the class documentation.

**Reply by waszee** [reply]
Following are some of the streams I use.  I am in South Carolina but these stations are from Boston Mass.  The Wgbh stations are playing poorly. They have icy-metaint:4096.  The WUMB stations generally play well and they are using icy-metaint:16000,  WDAV is closer in Charlotte, NC.  I also have those Swiss stations that you use in my list.  They play well sometimes depending on time of day.

  "https://wgbh-live.streamguys1.com/wgbh",
  "https://wgbh-live.streamguys1.com/classical-hi/",
  "https://wgbh-live.streamguys1.com/BSOConcert-8106",
  "https://wgbh-live.streamguys1.com/BostonEarlyMusic-8112",
  "https://wgbh-live.streamguys1.com/Bach-8108",
  "https://wumb.streamguys1.com/wumb919fast",
  "https://wumb.streamguys1.com/wumb919cf", // contemp folk
  "https://wumb.streamguys1.com/wumb919hol",// holiday
  "https://wumb.streamguys1.com/wumb919fr",//french accent
  "https://wumb.streamguys1.com/wumb919ho2",//students
  "https://wumb.streamguys1.com/wumb919cj", // contemp celtic
  "https://audio-mp3.ibiblio.org/wdav-112k", // wdav

**Comment by waszee** [comment]
fyi - my current setup is simple ESP S3 WROOM1 w N16R8, a PCM 5102 module, and IR remote.  I am not using the SDMMC module on this test setup but have another version that puts the net login info, mp3 folders, and url lists into folders on a SD card using the fast mmc protocol and sends the metadata out to a 1306 digital display instead of the serial monitor.   Current setup seeks to improve station switch and use of the Ir remote.  I am sending the headphone out to an external amplifier but have also used the MAX98375 module to the logitech speaker with no problems.

![image](https://github.com/user-attachments/assets/744918fb-47ad-47ff-9283-02643b9bbeaa)

**Reply by waszee** [reply]
Ok I finally understood what "start".    I needed to look more closely at the player.begin.  I did not realize it set the index to the zero position by default and needed to override with actual desired value.  Instead of player.setIndex(dstation)  or modifying the urlStream   I just needed player.beign(dstation) and avoid using player.begin() which reset the stream to the initial index.   Code is much cleaner now. This appears to have  also cleaned up the metaint not defined error.  
'''
`  if (dstation!=thestation) {
    if (dstation<0){ dstation=maxstation-1;}
    if (dstation>=maxstation){ dstation=0;}
    player.end();
    player.begin(dstation);
    Serial.printf("station = %d  vol = %f \n", dstation,dvol);
    player.copy();
  }
'''
Thanks you for being patient.
